tests/integration/test_complex_ops.py

```diff
@@ -18,27 +18,82 @@
 
 import cunumeric as num
 
-DTYPES = [np.complex64, np.complex128]
+ARRAYS = (
+    [1, 2, 3],
+    [4j, 5j, 6j],
+    [3 + 6j],
+    [[1 + 4j, 2 + 5j, 3 + 6j]],
+    [],
+)
 
 
-@pytest.mark.parametrize("dtype", DTYPES)
-def test_array(dtype):
-    x_np = np.array([1 + 4j, 2 + 5j, 3 + 6j], dtype)
+def strict_type_equal_array(a, b):
+    return np.array_equal(a, b) and a.dtype == b.dtype
+
+
+def strict_type_equal_scalar(a, b):
+    return a == b and type(a) == type(b)
+
+
+@pytest.mark.parametrize("dtype", (np.complex64, np.complex128))
+@pytest.mark.parametrize("arr", ARRAYS)
+def test_complex_array(arr, dtype):
+    # If val has complex elements, the returned type is float.
+    x_np = np.array(arr, dtype)
     x_num = num.array(x_np)
 
-    assert num.array_equal(x_np.conj(), x_num.conj())
-    assert num.array_equal(x_np.real, x_num.real)
-    assert num.array_equal(x_np.imag, x_num.imag)
+    assert strict_type_equal_array(np.real(x_np), num.real(x_num))
+    assert strict_type_equal_array(np.imag(x_np), num.imag(x_num))
 
+    assert strict_type_equal_array(x_np.conj(), x_num.conj())
+    assert strict_type_equal_array(x_np.real, x_num.real)
+    assert strict_type_equal_array(x_np.imag, x_num.imag)
 
-@pytest.mark.parametrize("dtype", DTYPES)
-def test_single(dtype):
-    x_np = np.array([3 + 6j], dtype)
+
+@pytest.mark.parametrize("dtype", (np.int32, np.float64))
+def test_non_complex_array(dtype):
+    # If val is real, the type of val is used for the output.
+    arr = [1, 2, 3]
+    x_np = np.array(arr, dtype)
     x_num = num.array(x_np)
 
-    assert num.array_equal(x_np.conj(), x_num.conj())
-    assert num.array_equal(x_np.real, x_num.real)
-    assert num.array_equal(x_np.imag, x_num.imag)
+    assert strict_type_equal_array(np.real(x_np), num.real(x_num))
+    assert strict_type_equal_array(np.imag(x_np), num.imag(x_num))
+
+    assert strict_type_equal_array(x_np.conj(), x_num.conj())
+    assert strict_type_equal_array(x_np.real, x_num.real)
+    assert strict_type_equal_array(x_np.imag, x_num.imag)
+
+
+SCALARS = (1, 0.0, 1 + 1j, 1.1 + 1j, 0j)
+
+
+@pytest.mark.xfail
+@pytest.mark.parametrize("val", SCALARS)
+def test_scalar(val):
+    # e.g., strict_type_equal_scalar(1.1, array(1.1))
+    # In numpy, it returns val as a scalar
+    # In cunumeric, it returns a 0-dim array(val)
```

Comment:
@manopapad IIRC this is intentional or at least expected? If that's the case we can update the test to match the actual behaviour (and thus pass) but we might want to flag these cases in some way for easy reference.

Reply:
Note that `np.array(42)` and `42` are considered equal by `np.array_equal`. The reason the `strict_type_equal_scalar` check is failing here is that scalar values don't have a `.dtype` attribute.

Generally speaking scalars and 0d arrays are (supposed to be) interchangeable, from a duck typing standpoint. 
And NumPy's behavior when returning scalar values is not 100% consistent (sometimes it returns actual Python scalars, and other times 0d arrays). Given all the above, and the fact that eagerly returning a scalar requires us to block, we have generally stuck to returning 0d arrays, and letting the subsequent code block on the value if they need it right away (the value might instead be used in another NumPy operation, in which case we can just "string" those together, rather than block). So this behavior is indeed expected.

I agree that it would be good to differentiate between xfail's that mean "this is a bug, and once fixed this test should be passing" and those that mean "this is an expected deviation from NumPy behavior".

Reply:
I would actually propose changing the test to make it pass, leaving the comment, and then adding some custom mark like `pytest.mark.diff` or something.

Reply:
Fixed. Thanks.